### ISSUE TITLE
Regression: inferred return types in recursive functions are checked incorrectly

### ISSUE THREAD

**Issue by mkantor** [microsoft/TypeScript#62957]
### 🔎 Search Terms

inferred "return type" recursive error "false negative"

### 🕗 Version & Regression Information

- This changed between versions 5.7.3 and 5.8.3

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/PTAEFUDsGMHsFt4FNIBdQENIE9SwGaioAWSociKqSAJgLSwCu6q2ADmVpLKhqgJaxIAZyKxQwpJyLsySAE7zY8gHSgAggBtNoAFAhyCZGloNmmSN14ChojPLL8aVftAw7U4-pHwKHNGQ5RBzZlagCAI1wSJH0wTSwAc0YMRLJJeQA3VyQVXV04EXQkAA8MeDZNMgBeUAAKfAAueoBKUGqAPlBIRngIhRbm4VR5b0T2rvw6lpVPADF+Etpp-MLh0EJa6eAAKlAhkbHQHbBOonlGMgB+UAByK0xu3v75W-3QRJWDUB+APSvVrZ0OMti1du9hqNIOMThNQABvXQ-b7-fI-NboYjg5rTOGQo6w2r4JGgb5-AEkhyoRjySCgUrlSpIOrENoGABy4lYHHpimUAEISWTyboAL75T5ssAAFWISgA7nZQBc0PxkPSStAkGwbHS6gAiKYzeaLWigfiiB4YDaMGC6-UzIA

### 💻 Code

```ts
const example = (f: () => number): string => f().toFixed()

const f = () => true ? 'not a number' : g()

const g = () => {
  // There's no type error here, but `f` always returns strings, while `example`
  // expects a callback which returns numbers.
  return example(f)
}
```

### 🙁 Actual behavior

The above code is accepted by the type checker even though it contains a type error. This is a reduced reproduction of something I tripped over in a real codebase.

### 🙂 Expected behavior

I'd expect a type error; probably the `Argument of type '() => string' is not assignable to
parameter of type '() => number'` error which occurs in TypeScript 5.7.

### Additional information about the issue

As shown in [the linked playground](https://www.typescriptlang.org/play/?#code/PTAEFUDsGMHsFt4FNIBdQENIE9SwGaioAWSociKqSAJgLSwCu6q2ADmVpLKhqgJaxIAZyKxQwpJyLsySAE7zY8gHSgAggBtNoAFAhyCZGloNmmSN14ChojPLL8aVftAw7U4-pHwKHNGQ5RBzZlagCAI1wSJH0wTSwAc0YMRLJJeQA3VyQVXV04EXQkAA8MeDZNMgBeUAAKfAAueoBKUGqAPlBIRngIhRbm4VR5b0T2rvw6lpVPADF+Etpp-MLh0EJa6eAAKlAhkbHQHbBOonlGMgB+UAByK0xu3v75W-3QRJWDUB+APSvVrZ0OMti1du9hqNIOMThNQABvXQ-b7-fI-NboYjg5rTOGQo6w2r4JGgb5-AEkhyoRjySCgUrlSpIOrENoGABy4lYHHpimUAEISWTyboAL75T5ssAAFWISgA7nZQBc0PxkPSStAkGwbHS6gAiKYzeaLWigfiiB4YDaMGC6-UzIA), if explicit type annotations are added in various places then a type error does occur (even if the annotations are identical to the inferred types reported by the language service).

**Comment by RyanCavanaugh** [comment]
Bisects to #60208

**Comment by RyanCavanaugh** [comment]
What a mess. Repros in 6.0 nightly, but not in 7.0. The bug that the linked PR fixes doesn't repro in 7.0, but I'm not sure *why* 7.0 would have incidentally fixed the original bug but not regressed *this* test. I think we just want to call this "fixed in 7.0" rather than revert a PR that fixed a crash

**Comment by RyanCavanaugh** [comment]
cc @ahejlsberg in case you care. Kind of interesting.

**Comment by RyanCavanaugh** [comment]
This all started because of #3276 😵‍💫

**Comment by ahejlsberg** [comment]
It is indeed a mess. I couldn't make heads or tails of the call resolution error elaboration logic, so I reworked and simplified it as I ported the code (the revised code lives in the `reportCallResolutionErrors` function in `checker.go`). Apparently the simplification fixed the issue that was introduced in 5.7 and subsequently "fixed" in 5.8 by #60208. That logic isn't in 7.0 because the #60208 went in after the snapshot we ported. So, all up I'd say we're good for 7.0.
